src/testdrive/src/action/sql.rs

```diff
@@ -76,19 +79,24 @@ pub async fn run_sql(mut cmd: SqlCommand, state: &State) -> Result<ControlFlow,
 
     let query = &cmd.query;
     print_query(query, Some(&stmt));
-
-    let state = &state;
     let expected_output = &cmd.expected_output;
-    let res = match should_retry {
+    let (state, res) = match should_retry {
         true => Retry::default()
             .initial_backoff(state.initial_backoff)
             .factor(state.backoff_factor)
             .max_duration(state.timeout)
             .max_tries(state.max_tries),
         false => Retry::default().max_duration(state.timeout).max_tries(1),
     }
-    .retry_async_canceling(|retry_state| async move {
-        match try_run_sql(state, query, expected_output).await {
+    .retry_async_with_state(state, |retry_state, state| async move {
+        match try_run_sql(
+            state,
+            query,
+            expected_output,
+            retry_state.i + 1 < state.max_tries && should_retry,
```

Comment:
It might be nice to pull this up into a variable, e.g.
```
let should_continue = retry_state.i + 1 < state.max_tries && should_retry
```
as-is I'm not totally sure what this value does